test/integration/xds_integration_test.cc

```diff
@@ -568,5 +580,60 @@ TEST_P(LdsIntegrationTest, FailConfigLoad) {
   });
   EXPECT_DEATH(initialize(), "Didn't find a registered implementation for name: 'grewgragra'");
 }
+
+// Verify that the listener in place update will accomplish anyway if the listener is removed.
+TEST_P(LdsInplaceUpdateTcpProxyIntegrationTest,
+       TcpListenerRemoveFilterChainCalledAfterListenerIsRemoved) {
+  // The in place listener update takes 2 seconds. We will remove the listener.
+  drain_time_ = std::chrono::seconds(2);
+  // 1. Start the first in place listener update.
+  setUpstreamCount(2);
+  initialize();
+  std::string response_0;
+  auto client_conn_0 = createConnectionAndWrite("alpn0", "hello", response_0);
+  client_conn_0->waitForConnection();
+  FakeRawConnectionPtr fake_upstream_connection_0;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection_0));
+
+  ConfigHelper new_config_helper(
+      version_, *api_, MessageUtil::getJsonStringFromMessageOrDie(config_helper_.bootstrap()));
+  new_config_helper.addConfigModifier(
+      [&](envoy::config::bootstrap::v3::Bootstrap& bootstrap) -> void {
+        auto* listener = bootstrap.mutable_static_resources()->mutable_listeners(0);
+        listener->mutable_filter_chains()->RemoveLast();
+      });
+  new_config_helper.setLds("1");
+
+  // 2. Remove the tcp listener. At some extreme cases this removal will be stacked with the the
```

Comment:
What is meant by extreme cases? Are you relying on racy timing behavior?

Reply:
Yeah, in the integration test there is no guaranteed way to reproduce racy timer expiration.